### PR TITLE
small fixes in API comments. first draft of README file

### PR DESCRIPTION
**Pull request by dkozorez**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# OAuth2 device flow library
+
+This library provides an easy way to use Googlee Firebase authorization.
```
Comment (ppetrosh) [inline README.md]: Google

File: README.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+Basic functionality sonsists of following actions:
```
Comment (ppetrosh) [inline README.md]: consists

File: README.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
+# OAuth2 device flow library
+
+This library provides an easy way to use Google Firebase authorization.
+
+Basic functionality consists of following actions:
+
+- Asynchronous token acquisition
+- Synchronous token provision if it is valid
```
Comment (ppetrosh) [inline README.md]: What does it mean?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+- Checking if token is valid
+- Checking if the application is authorized
+- Refresh operation for expired access token
+
+**To add this library to your project, add** `#require "OAuth2.DeviceFlow.agent.lib.nut:1.0.0"` **to the top of your agent code**
+
+
+## OAuth2.DeviceFlow.agent public methods
```
Comment (ppetrosh) [inline README.md]: `OAuth2.DeviceFlow.agent`

File: README.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+### constructor(provider, params)
+
+Client constructor. Accepts service specific OAuth2 provider configuration table `provider` and user specific table `params`.
+The first argument must contain `LOGIN_HOST`, `TOKEN_HOST` may contain optional `GRANT_TYPE`. The second 
+argument must contain `clientId`, `scope` and may contain optional `clientSecret`.
```
Comment (ppetrosh) [inline README.md]: It feels like there should be two table each describing the content of `provider` and `params` tables. I would also rename them to `providerSettings` and `userSettings` respectively. Please look at https://github.com/electricimp/messagemanager for example.

Each table should contain:

| Key | Data Type | Description |
| ----- | -------------- | --------------- |
| | | |

File: README.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+### constructor(provider, params)
+
+Client constructor. Accepts service specific OAuth2 provider configuration table `provider` and user specific table `params`.
+The first argument must contain `LOGIN_HOST`, `TOKEN_HOST` may contain optional `GRANT_TYPE`. The second 
+argument must contain `clientId`, `scope` and may contain optional `clientSecret`. 
+
+#### Example
+
+```squirrel
+    GOOGLE =  {
```
Comment (ppetrosh) [inline README.md]: ->`providerSettings `

File: README.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        "LOGIN_HOST" : "https://accounts.google.com/o/oauth2/device/code",
+        "TOKEN_HOST" : "https://www.googleapis.com/oauth2/v4/token",
+        "GRANT_TYPE" : "http://oauth.net/grant_type/device/1.0",
+    };
+    local data = {
```
Comment (ppetrosh) [inline README.md]: ->`userSettings`

File: README.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+```
+
+### acquireAccessToken(tokenReadyCallback, notifyUserCallback, force)
+
+Starts access token acquisition procedure. Depending on Client state may starts full client authorization procedure or
+just token refreshing. 
+
+Parameter details:
+
+| Parameter | Description |
```
Comment (ppetrosh) [inline README.md]: I would add the second column: `Type`

File: README.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+| --- | --- |
+| *tokenReadyCallback* | The handler to be called when access token is acquired or error is observed |
+| *notifyUserCallback* | The handler to be called when user action is required. See [RFE, device flow, section3.3](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-05#section-3.3) |
+| *[force]* | [optional] the directive to start new acquisition procedure even if previous request is not complete. False by default |
+
+`tokenReadyCallback` should accept two parameters: first is an access token string, the second is an error description string if something went wrong during the acquisition procedure.
```
Comment (ppetrosh) [inline README.md]: Each of the callbacks need to have examples (or just one that combines both).

Comment (ppetrosh) [inline README.md]: It would also be nice to have the arguments described as tables.

File: README.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+Checks if access token is valid.
+
+### isAuthorized()
+
+Checks if the client is authorized and able to refresh expired access token.
+
+
+### refreshAccessToken(tokenReadyCallback)
+
+Refreshes access token non blocking way, will invoke `tokenReadyCallback` in case of success.
```
Comment (ppetrosh) [inline README.md]: It would also be nice to have the arguments described in a table (with name, type description columns).

**Comment by ppetrosh** [comment]
A general comment:

There should be 
1. Source code examples for every API call
2. One complete example of using the library.

File: README.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
+# OAuth2 device flow library
+
+This library provides an easy way to use Google Firebase authorization.
```
Comment (sergey-nbl) [inline README.md]: Browserless and limited device authorization. Google is only exmaple

File: README.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+**To add this library to your project, add** `#require "OAuth2.DeviceFlow.agent.lib.nut:1.0.0"` **to the top of your agent code**
+
+
+## OAuth2.DeviceFlow.agent public methods
+
+### constructor(provider, params)
+
+Client constructor. Accepts service specific OAuth2 provider configuration table `provider` and user specific table `params`.
+The first argument must contain `LOGIN_HOST`, `TOKEN_HOST` may contain optional `GRANT_TYPE`. The second
```
Comment (sergey-nbl) [inline README.md]: Need few word about optional GRANT_TYPE is set to OAUTH2_DEVICE_FLOW_GRANT_TYPE value by default.   All other values are just not used if empty

File: README.md (hunk 11 above)
Comment (sergey-nbl) [inline README.md]: Need link to the RFE

File: README.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
+# OAuth2 device flow library
```
Comment (sergey-nbl) [inline README.md]: Library to support OAUth2 authorization for browserless and input constrained devices

File: README.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+Basic functionality consists of following actions:
+
+- Asynchronous token acquisition
+- Synchronous token provision if it is valid
+- Checking if token is valid
+- Checking if the application is authorized
+- Refresh operation for expired access token
+
+**To add this library to your project, add** `#require "OAuth2.DeviceFlow.agent.lib.nut:1.0.0"` **to the top of your agent code**
```
Comment (sergey-nbl) [inline README.md]: Need more words about predefined config GOOGLE

File: README.md (hunk 3 above)
Comment (dkozorez) [inline README.md]: Will replace with
- Token acquisition

File: README.md (hunk 4 above)
Comment (dkozorez) [inline README.md]: fixed

File: README.md (hunk 5 above)
Comment (dkozorez) [inline README.md]: fixed

File: README.md (hunk 6 above)
Comment (dkozorez) [inline README.md]: fixed

File: README.md (hunk 7 above)
Comment (dkozorez) [inline README.md]: fixed

File: README.md (hunk 8 above)
Comment (dkozorez) [inline README.md]: fixed

File: README.md (hunk 9 above)
Comment (dkozorez) [inline README.md]: fixed

File: README.md (hunk 10 above)
Comment (dkozorez) [inline README.md]: fixed

File: README.md (hunk 11 above)
Comment (dkozorez) [inline README.md]: fixed
з.ы. откуда ты это сказал? ты же в отпуске

File: README.md (hunk 13 above)
Comment (dkozorez) [inline README.md]: fixed

File: README.md (hunk 12 above)
Comment (dkozorez) [inline README.md]: fixed

File: README.md (hunk 14 above)
Comment (dkozorez) [inline README.md]: fixed

File: README.md [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+- Refresh operation for expired access token
+
+As one of possible usages library  provides table named `GOOGLE` that contains constants specific to Google Firebase use case. The table 
+provides both `LOGIN_HOST` and `TOKEN_HOST` and also overrides `GRANT_TYPE`. The one specified by RFE does not fit Firebase case.
+If user leaves it empty, default `urn:ietf:params:oauth:grant-type:device_code` will be used.
+
+**To add this library to your project, add** `#require "OAuth2.DeviceFlow.Client.nut:1.0.0"` **to the top of your agent code**
+
+
+## OAuth2.DeviceFlow.agent public methods
```
Comment (sergey-nbl) [inline README.md]: DeviceFlow.Client

File: README.md [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+
+### constructor(provider, params)
+
+Client constructor. Accepts service specific OAuth2 provider configuration table `provider` and user specific table `params`.
```
Comment (sergey-nbl) [inline README.md]: provider and params not renamed

File: README.md [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+        if (err) {
+            server.error(err);
+        }
+    }
+);
+```
+
+## Complete usage sample
+
+To connect all the parts together and show a sample of common case of library usage let's take a look a following sample
```
Comment (sergey-nbl) [inline README.md]: Need more words what we are doing and why. The doesn't look simple enough. What if we reuse test and exclude token verification?

File: README.md (hunk 15 above)
Comment (ppetrosh) [inline README.md]: 👍 I remember myself making the same comment awhile ago already

File: README.md [hunk 18]
```diff
@@ -0,0 +1,235 @@
+# Library to support OAUth2 authorization for browserless and input constrained devices
```
Comment (ppetrosh) [inline README.md]: 1. `OAUth2` -> `OAuth2`
2. It should be just `OAUth2` in the title and the description following it.

File: README.md [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+- Token acquisition
+- Checking if token is valid
+- Checking if the application is authorized
+- Refresh operation for expired access token
+
+As one of possible usages library  provides table named `GOOGLE` that contains constants specific to Google Firebase use case. The table 
+provides both `LOGIN_HOST` and `TOKEN_HOST` and also overrides `GRANT_TYPE`. The one specified by RFE does not fit Firebase case.
+If user leaves it empty, default `urn:ietf:params:oauth:grant-type:device_code` will be used.
+
+**To add this library to your project, add** `#require "OAuth2.DeviceFlow.Client.nut:1.0.0"` **to the top of your agent code**
```
Comment (ppetrosh) [inline README.md]: The library name is `OAuth2.DeviceFlow.agent.lib.nut`

File: README.md [hunk 20]
```diff
@@ -0,0 +1,235 @@
+# Library to support OAUth2 authorization for browserless and input constrained devices
+
+This library provides an easy way to use browserless and [limited device authorization](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-05).
```
Comment (ppetrosh) [inline README.md]: In fact this has never been a requirement to support the spec, which is not released yet.
It's just that at some point it was discovered that what works for Google Firebase looks very similar to the specification.
So I would just mention this fact at the end of the description.

The goal of the library is to support generic oauth flow (specifically Firebase).

File: README.md [hunk 21]
```diff
@@ -0,0 +1,235 @@
...
+    }
+);
+```
+
+## Complete usage sample
+
+To connect all the parts together and show a sample of common case of library usage let's take a look a following sample
+
+```squirrel
+//fill CLIENT_ID and CLIENT_SECRET with correct values
```
Comment (ppetrosh) [inline README.md]: The sample here should be at least two times shorter/simpler if possible. It should be just a reference.

File: README.md (hunk 15 above)
Comment (dkozorez) [inline README.md]: Applied to the wrong line. Fixed with renaming of sub-title.

File: README.md (hunk 16 above)
Comment (dkozorez) [inline README.md]: fixed

File: README.md (hunk 18 above)
Comment (dkozorez) [inline README.md]: Fixed 1, not sure about 2. Haven't find any misspelling of OAuth2 except the title.

File: README.md (hunk 19 above)
Comment (dkozorez) [inline README.md]: fixed

File: README.md (hunk 20 above)
Comment (dkozorez) [inline README.md]: Didn't get the idea -- if the goal is to support  general case of OAuth2 device flow, than title looks ok. Did I  put the stress on incorrect place?

File: README.md (hunk 21 above)
Comment (dkozorez) [inline README.md]: Part of the sample regarding Rocky handle for /token could be easily removed without any real damage. Not sure if it's a good idea to remove part about token refresh.

File: README.md [hunk 22]
```diff
@@ -0,0 +1,197 @@
+# OAuth2 authorization for browserless and input constrained devices
```
Comment (ppetrosh) [inline README.md]: I would change it to:
```
# OAuth2 DeviceFlow Client

This library implements the OAuth2 authorization flow for for browserless and input constrained devices. The implementation conforms to limited device authorization [spec](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-05).

The main library functionality consists of following actions:
...
```

**Pull request merged**